{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1: KEGG and gene id mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Familiarize yourself with the KEGG Rest interface and how to access it with Biopyhton:\n",
    "\n",
    "http://www.genome.jp/kegg/rest/keggapi.html\n",
    "\n",
    "http://nbviewer.jupyter.org/github/widdowquinn/notebooks/blob/master/Biopython_KGML_intro.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 1.1 Extract gene lists for all (mouse) KEGG pathways and store them in a suitable Python data structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.68'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import Bio\n",
    "Bio.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.KEGG.REST import *\n",
    "from Bio.KEGG.KGML import KGML_parser\n",
    "from Bio.Graphics.KGML_vis import KGMLCanvas\n",
    "from Bio.KEGG import REST\n",
    "from Bio.Graphics.ColorSpiral import ColorSpiral\n",
    "from IPython.display import Image, HTML\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.KEGG import REST\n",
    "import math\n",
    "import re\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as sp\n",
    "# A bit of code that will help us display the PDF output\n",
    "def PDF(filename):\n",
    "    return HTML('<iframe src=%s width=700 height=350></iframe>' % filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-101-24f6d0deb9c3>, line 5)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-101-24f6d0deb9c3>\"\u001b[0;36m, line \u001b[0;32m5\u001b[0m\n\u001b[0;31m    print \"\\n\".join(text.split('\\n')[:lines] + ['[...]'])\u001b[0m\n\u001b[0m             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# A bit of helper code to shorten long text\n",
    "def head(text, lines=10):\n",
    "    \"\"\" Print the first lines lines of the passed text.\n",
    "    \"\"\"\n",
    "    print \"\\n\".join(text.split('\\n')[:lines] + ['[...]'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mouse_pathways = REST.kegg_list(\"pathway\", \"mmu\").read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#mouse_pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "HTTPError",
     "evalue": "HTTP Error 400: Bad Request",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mHTTPError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-131-5df3215863c0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mmouse_genes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mpathway\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpathways\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mpathway_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mREST\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkegg_get\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpathway\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# query and read each pathway\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0;31m# iterate through each KEGG pathway file, keeping track of which section\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;31m# of the file we're in, only read the gene in each pathway\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/site-packages/Bio/KEGG/REST.py\u001b[0m in \u001b[0;36mkegg_get\u001b[0;34m(dbentries, option)\u001b[0m\n\u001b[1;32m    196\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid option arg for kegg get request.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    197\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 198\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_q\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"get\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdbentries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    199\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    200\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/site-packages/Bio/KEGG/REST.py\u001b[0m in \u001b[0;36m_q\u001b[0;34m(op, arg1, arg2, arg3)\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m         \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"%s/%s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 42\u001b[0;31m     \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_urlopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mURL\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     43\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m\"image\"\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0marg2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/urllib/request.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    161\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    162\u001b[0m         \u001b[0mopener\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_opener\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 163\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mopener\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    164\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0minstall_opener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopener\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    470\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mprocessor\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprocess_response\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    471\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 472\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    473\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    474\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    580\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    581\u001b[0m             response = self.parent.error(\n\u001b[0;32m--> 582\u001b[0;31m                 'http', request, response, code, msg, hdrs)\n\u001b[0m\u001b[1;32m    583\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    584\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    508\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mhttp_err\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    509\u001b[0m             \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'default'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'http_error_default'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0morig_args\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 510\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    511\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    512\u001b[0m \u001b[0;31m# XXX probably also want an abstract factory that knows when it makes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    442\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    443\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 444\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    445\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    446\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Timo/anaconda3/lib/python3.5/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_default\u001b[0;34m(self, req, fp, code, msg, hdrs)\u001b[0m\n\u001b[1;32m    588\u001b[0m \u001b[0;32mclass\u001b[0m \u001b[0mHTTPDefaultErrorHandler\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mBaseHandler\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    589\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mhttp_error_default\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 590\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mHTTPError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_url\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    591\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    592\u001b[0m \u001b[0;32mclass\u001b[0m \u001b[0mHTTPRedirectHandler\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mBaseHandler\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mHTTPError\u001b[0m: HTTP Error 400: Bad Request"
     ]
    }
   ],
   "source": [
    "# Get the genes for pathways and add them to a list\n",
    "mouse_genes = [] \n",
    "for pathway in pathways:\n",
    "    pathway_file = REST.kegg_get(pathway).read()  # query and read each pathway\n",
    "    # iterate through each KEGG pathway file, keeping track of which section\n",
    "    # of the file we're in, only read the gene in each pathway\n",
    "    current_section = None\n",
    "    for line in pathway_file.rstrip().split(\"\\n\"):\n",
    "        section = line[:12].strip()  # section names are within 12 columns\n",
    "        if not section == \"\":\n",
    "            current_section = section\n",
    "        if current_section == \"GENE\":\n",
    "            if len(line[12:].split(\"; \")) > 1:\n",
    "                gene_identifiers, gene_description = line[12:].split(\"; \")\n",
    "            else:\n",
    "                continue\n",
    "            if len(gene_identifiers) == 2:\n",
    "                gene_id, gene_symbol = gene_identifiers.split()\n",
    "            else:\n",
    "                continue\n",
    "            if not gene_symbol in mouse_genes:\n",
    "                mouse_genes.append(gene_symbol)\n",
    "print(\"There are %d pathways and %d genes. The genes are:\" % \\\n",
    "      (len(pathways), len(mouse_genes)))\n",
    "print(\", \".join(mouse_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "file = open('mmu_genes.txt','r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "text = file.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'path:mmu00010\\tmmu:100042025\\n'"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "length = len(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pathways = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i in xrange(1,length):\n",
    "    pathway,gene = text[i].split()\n",
    "    gene = gene.split(\"mmu:\")[1]\n",
    "    pathway = pathway.split(\":mmu\")[1]\n",
    "    pathways.append(pathway)\n",
    "    genes.append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['103988',\n",
       " '106557',\n",
       " '110695',\n",
       " '11522',\n",
       " '11529',\n",
       " '11532',\n",
       " '11669',\n",
       " '11670',\n",
       " '11671',\n",
       " '11674',\n",
       " '11676',\n",
       " '12183',\n",
       " '13382',\n",
       " '13806',\n",
       " '13807',\n",
       " '13808',\n",
       " '14120',\n",
       " '14121',\n",
       " '14377',\n",
       " '14378',\n",
       " '14433',\n",
       " '14447',\n",
       " '14751',\n",
       " '15275',\n",
       " '15277',\n",
       " '16828',\n",
       " '16832',\n",
       " '16833',\n",
       " '17330',\n",
       " '18534',\n",
       " '18597',\n",
       " '18598',\n",
       " '18641',\n",
       " '18642',\n",
       " '18648',\n",
       " '18655',\n",
       " '18663',\n",
       " '18746',\n",
       " '18770',\n",
       " '212032',\n",
       " '216019',\n",
       " '21991',\n",
       " '230163',\n",
       " '235339',\n",
       " '26876',\n",
       " '319625',\n",
       " '353204',\n",
       " '433182',\n",
       " '56012',\n",
       " '56421',\n",
       " '56752',\n",
       " '56847',\n",
       " '58810',\n",
       " '60525',\n",
       " '621603',\n",
       " '66681',\n",
       " '67689',\n",
       " '68263',\n",
       " '68401',\n",
       " '68738',\n",
       " '72141',\n",
       " '72157',\n",
       " '72535',\n",
       " '73458',\n",
       " '74551',\n",
       " '104112',\n",
       " '11428',\n",
       " '11429',\n",
       " '12974',\n",
       " '13382',\n",
       " '14194',\n",
       " '15926',\n",
       " '15929',\n",
       " '170718',\n",
       " '17448',\n",
       " '17449',\n",
       " '18293',\n",
       " '18534',\n",
       " '18563',\n",
       " '18597',\n",
       " '18598',\n",
       " '20916',\n",
       " '20917',\n",
       " '235339',\n",
       " '239017',\n",
       " '243996',\n",
       " '269951',\n",
       " '56451',\n",
       " '66052',\n",
       " '66925',\n",
       " '66945',\n",
       " '67680',\n",
       " '67834',\n",
       " '68263',\n",
       " '71832',\n",
       " '74551',\n",
       " '78920',\n",
       " '100198',\n",
       " '110208',\n",
       " '110639',\n",
       " '11674',\n",
       " '11676',\n",
       " '14120',\n",
       " '14121',\n",
       " '14380',\n",
       " '14381',\n",
       " '14751',\n",
       " '18641',\n",
       " '18642',\n",
       " '19139',\n",
       " '19733',\n",
       " '19895',\n",
       " '21351',\n",
       " '21881',\n",
       " '230163',\n",
       " '232449',\n",
       " '235582',\n",
       " '328099',\n",
       " '353204',\n",
       " '56421',\n",
       " '66171',\n",
       " '66646',\n",
       " '66681',\n",
       " '71336',\n",
       " '72157',\n",
       " '74419',\n",
       " '75456',\n",
       " '75731',\n",
       " '83553',\n",
       " '100559',\n",
       " '100727',\n",
       " '102448',\n",
       " '110006',\n",
       " '112417',\n",
       " '11669',\n",
       " '11671',\n",
       " '11677',\n",
       " '11997',\n",
       " '14187',\n",
       " '16591',\n",
       " '20322',\n",
       " '216558',\n",
       " '22235',\n",
       " '22236',\n",
       " '22238',\n",
       " '231396',\n",
       " '243085',\n",
       " '394430',\n",
       " '394432',\n",
       " '394433',\n",
       " '394434',\n",
       " '394435',\n",
       " '394436',\n",
       " '552899',\n",
       " '58810',\n",
       " '66646',\n",
       " '67861',\n",
       " '67880',\n",
       " '68631',\n",
       " '71755',\n",
       " '71773',\n",
       " '72094',\n",
       " '72535',\n",
       " '94215',\n",
       " '94284',\n",
       " '110119',\n",
       " '11674',\n",
       " '11676',\n",
       " '11677',\n",
       " '11997',\n",
       " '14120',\n",
       " '14121',\n",
       " '14187',\n",
       " '15275',\n",
       " '15277',\n",
       " '16548',\n",
       " '170768',\n",
       " '18639',\n",
       " '18640',\n",
       " '18641',\n",
       " '18642',\n",
       " '20322',\n",
       " '212032',\n",
       " '216019',\n",
       " '218138',\n",
       " '21991',\n",
       " '22122',\n",
       " '225913',\n",
       " '230163',\n",
       " '234730',\n",
       " '270198',\n",
       " '29858',\n",
       " '319801',\n",
       " '331026',\n",
       " '353204',\n",
       " '54128',\n",
       " '56421',\n",
       " '67861',\n",
       " '69080',\n",
       " '75540',\n",
       " '103988',\n",
       " '11605',\n",
       " '11677',\n",
       " '11997',\n",
       " '12091',\n",
       " '14187',\n",
       " '14377',\n",
       " '14378',\n",
       " '14387',\n",
       " '14430',\n",
       " '14595',\n",
       " '14635',\n",
       " '15275',\n",
       " '15277',\n",
       " '16770',\n",
       " '18641',\n",
       " '18642',\n",
       " '212032',\n",
       " '216019',\n",
       " '216558',\n",
       " '226413',\n",
       " '232714',\n",
       " '319625',\n",
       " '53418',\n",
       " '56421',\n",
       " '66681',\n",
       " '67861',\n",
       " '68401',\n",
       " '69983',\n",
       " '72157',\n",
       " '74246',\n",
       " '76051',\n",
       " '100559',\n",
       " '100727',\n",
       " '110695',\n",
       " '112417',\n",
       " '11669',\n",
       " '11671',\n",
       " '19733',\n",
       " '22235',\n",
       " '22236',\n",
       " '22238',\n",
       " '231396',\n",
       " '243085',\n",
       " '268756',\n",
       " '394430',\n",
       " '394432',\n",
       " '394433',\n",
       " '394434',\n",
       " '394435',\n",
       " '394436',\n",
       " '552899',\n",
       " '56727',\n",
       " '56752',\n",
       " '71773',\n",
       " '72094',\n",
       " '72535',\n",
       " '94215',\n",
       " '94284',\n",
       " '100705',\n",
       " '107476',\n",
       " '14081',\n",
       " '14104',\n",
       " '216739',\n",
       " '223722',\n",
       " '328845',\n",
       " '433256',\n",
       " '50790',\n",
       " '71147',\n",
       " '74205',\n",
       " '78625',\n",
       " '94180',\n",
       " '99035',\n",
       " '106529',\n",
       " '12686',\n",
       " '15107',\n",
       " '170439',\n",
       " '171210',\n",
       " '171281',\n",
       " '171282',\n",
       " '19063',\n",
       " '217698',\n",
       " '231086',\n",
       " '26897',\n",
       " '26922',\n",
       " '30963',\n",
       " '52538',\n",
       " '54325',\n",
       " '54326',\n",
       " '54397',\n",
       " '56348',\n",
       " '57874',\n",
       " '66775',\n",
       " '68801',\n",
       " '70025',\n",
       " '70757',\n",
       " '74559',\n",
       " '83603',\n",
       " '93747',\n",
       " '97212',\n",
       " '100040843',\n",
       " '110446',\n",
       " '110460',\n",
       " '110695',\n",
       " '11363',\n",
       " '11364',\n",
       " '11370',\n",
       " '113868',\n",
       " '11409',\n",
       " '11430',\n",
       " '11522',\n",
       " '11529',\n",
       " '11532',\n",
       " '11669',\n",
       " '11671',\n",
       " '12894',\n",
       " '12895',\n",
       " '12896',\n",
       " '13117',\n",
       " '13118',\n",
       " '13119',\n",
       " '13177',\n",
       " '14081',\n",
       " '15107',\n",
       " '216739',\n",
       " '230639',\n",
       " '231086',\n",
       " '235674',\n",
       " '23986',\n",
       " '26876',\n",
       " '270076',\n",
       " '277753',\n",
       " '328845',\n",
       " '433256',\n",
       " '435802',\n",
       " '50790',\n",
       " '52538',\n",
       " '56752',\n",
       " '666168',\n",
       " '66885',\n",
       " '72535',\n",
       " '74147',\n",
       " '74205',\n",
       " '78070',\n",
       " '78625',\n",
       " '80911',\n",
       " '93747',\n",
       " '94180',\n",
       " '97212',\n",
       " '110446',\n",
       " '110460',\n",
       " '15356',\n",
       " '15360',\n",
       " '208715',\n",
       " '208982',\n",
       " '353371',\n",
       " '64059',\n",
       " '67041',\n",
       " '69772',\n",
       " '71911',\n",
       " '12613',\n",
       " '13081',\n",
       " '13115',\n",
       " '13121',\n",
       " '13360',\n",
       " '13595',\n",
       " '14137',\n",
       " '15490',\n",
       " '16889',\n",
       " '16987',\n",
       " '18194',\n",
       " '20652',\n",
       " '20775',\n",
       " '223920',\n",
       " '235293',\n",
       " '244209',\n",
       " '66234',\n",
       " '73166',\n",
       " '74754',\n",
       " '101502',\n",
       " '104086',\n",
       " '12012',\n",
       " '12642',\n",
       " '13116',\n",
       " '13122',\n",
       " '13123',\n",
       " '13124',\n",
       " '15488',\n",
       " '170789',\n",
       " '17117',\n",
       " '20280',\n",
       " '208665',\n",
       " '26459',\n",
       " '56050',\n",
       " '93732',\n",
       " '12850',\n",
       " '15445',\n",
       " '18104',\n",
       " '217707',\n",
       " '230027',\n",
       " '234724',\n",
       " '27973',\n",
       " '52064',\n",
       " '56316',\n",
       " '69568',\n",
       " '71883',\n",
       " '100041375',\n",
       " '100043456',\n",
       " '100559',\n",
       " '100727',\n",
       " '105349',\n",
       " '107141',\n",
       " '110115',\n",
       " '112417',\n",
       " '12846',\n",
       " '13070',\n",
       " '13072',\n",
       " '13074',\n",
       " '13075',\n",
       " '13076',\n",
       " '13077',\n",
       " '13078',\n",
       " '13079',\n",
       " '13088',\n",
       " '13089',\n",
       " '13090',\n",
       " '13094',\n",
       " '13095',\n",
       " '13096',\n",
       " '13097',\n",
       " '13098',\n",
       " '13099',\n",
       " '13101',\n",
       " '13105',\n",
       " '13106',\n",
       " '13112',\n",
       " '13113',\n",
       " '13114',\n",
       " '13122',\n",
       " '13123',\n",
       " '14979',\n",
       " '15483',\n",
       " '15484',\n",
       " '15485',\n",
       " '15486',\n",
       " '15487',\n",
       " '15490',\n",
       " '15492',\n",
       " '15493',\n",
       " '15495',\n",
       " '15496',\n",
       " '15497',\n",
       " '20860',\n",
       " '208665',\n",
       " '20905',\n",
       " '22236',\n",
       " '22238',\n",
       " '223706',\n",
       " '226105',\n",
       " '226143',\n",
       " '231396',\n",
       " '243085',\n",
       " '243881',\n",
       " '27400',\n",
       " '337924',\n",
       " '380997',\n",
       " '394430',\n",
       " '394432',\n",
       " '394433',\n",
       " '394434',\n",
       " '394435',\n",
       " '394436',\n",
       " '404195',\n",
       " '433247',\n",
       " '53973',\n",
       " '54200',\n",
       " '545123',\n",
       " '552899',\n",
       " '56348',\n",
       " '56388',\n",
       " '56448',\n",
       " '57357',\n",
       " '69888',\n",
       " '71754',\n",
       " '71773',\n",
       " '72082',\n",
       " '72094',\n",
       " '72303',\n",
       " '76279',\n",
       " '78925',\n",
       " '94215',\n",
       " '94224',\n",
       " '94284',\n",
       " '100041273',\n",
       " '100042503',\n",
       " '104130',\n",
       " '108664',\n",
       " '110323',\n",
       " '110935',\n",
       " '114143',\n",
       " '11944',\n",
       " '11945',\n",
       " '11946',\n",
       " '11947',\n",
       " '11949',\n",
       " '11950',\n",
       " '11951',\n",
       " '11957',\n",
       " '11958',\n",
       " '11964',\n",
       " '11966',\n",
       " '11972',\n",
       " '11973',\n",
       " '11974',\n",
       " '11975',\n",
       " '11984',\n",
       " '12856',\n",
       " '12857',\n",
       " '12858',\n",
       " '12859',\n",
       " '12861',\n",
       " '12862',\n",
       " '12864',\n",
       " '12865',\n",
       " '12866',\n",
       " '12867',\n",
       " '12868',\n",
       " '12869',\n",
       " '140494',\n",
       " '17705',\n",
       " '17706',\n",
       " '17708',\n",
       " '17709',\n",
       " '17710',\n",
       " '17711',\n",
       " '17716',\n",
       " '17717',\n",
       " '17718',\n",
       " '17719',\n",
       " '17720',\n",
       " '17721',\n",
       " '17722',\n",
       " '17991',\n",
       " '17992',\n",
       " '17993',\n",
       " '17995',\n",
       " '192113',\n",
       " '20463',\n",
       " '21871',\n",
       " '22272',\n",
       " '22273',\n",
       " '225887',\n",
       " '226139',\n",
       " '226646',\n",
       " '227197',\n",
       " '228033',\n",
       " '230075',\n",
       " '239760',\n",
       " '242341',\n",
       " '27060',\n",
       " '27425',\n",
       " '28080',\n",
       " '333182',\n",
       " '338375',\n",
       " '407785',\n",
       " '407790',\n",
       " '54405',\n",
       " '54411',\n",
       " '57423',\n",
       " '595136',\n",
       " '66043',\n",
       " '66046',\n",
       " '66052',\n",
       " '66091',\n",
       " '66108',\n",
       " '66142',\n",
       " '66144',\n",
       " '66152',\n",
       " '66218',\n",
       " '66237',\n",
       " '66290',\n",
       " '66335',\n",
       " '66377',\n",
       " '66414',\n",
       " '66416',\n",
       " '66445',\n",
       " '66495',\n",
       " '66576',\n",
       " '66594',\n",
       " '66694',\n",
       " '66916',\n",
       " '66925',\n",
       " '66945',\n",
       " '67003',\n",
       " '67126',\n",
       " '67130',\n",
       " '67184',\n",
       " '67264',\n",
       " '67273',\n",
       " '67530',\n",
       " '67680',\n",
       " '67895',\n",
       " '67942',\n",
       " '68194',\n",
       " '68197',\n",
       " '68198',\n",
       " '68202',\n",
       " '68342',\n",
       " '68349',\n",
       " '68375',\n",
       " '68775',\n",
       " '69802',\n",
       " '69875',\n",
       " '70316',\n",
       " '70383',\n",
       " '71679',\n",
       " '72900',\n",
       " '73834',\n",
       " '74776',\n",
       " '74915',\n",
       " '75406',\n",
       " '75483',\n",
       " '76252',\n",
       " '76429',\n",
       " '78174',\n",
       " '78330',\n",
       " '84682',\n",
       " '108682',\n",
       " '109652',\n",
       " '109900',\n",
       " '11846',\n",
       " '11847',\n",
       " '11898',\n",
       " '14645',\n",
       " '14660',\n",
       " '14661',\n",
       " '14718',\n",
       " '14719',\n",
       " '18125',\n",
       " '18126',\n",
       " '18127',\n",
       " '18416',\n",
       " '216456',\n",
       " '217214',\n",
       " '227231',\n",
       " '76282',\n",
       " '100862375',\n",
       " '100862456',\n",
       " '102216272',\n",
       " '102641086',\n",
       " '103850',\n",
       " '104110',\n",
       " '104111',\n",
       " '105446',\n",
       " '107569',\n",
       " '108147',\n",
       " '109674',\n",
       " '110385',\n",
       " '110639',\n",
       " '110855',\n",
       " '11486',\n",
       " '11512',\n",
       " '11513',\n",
       " '11514',\n",
       " '11515',\n",
       " '11534',\n",
       " '11564',\n",
       " '11565',\n",
       " '11566',\n",
       " '11636',\n",
       " '11637',\n",
       " '11639',\n",
       " '11717',\n",
       " '11821',\n",
       " '12495',\n",
       " '12496',\n",
       " '12497',\n",
       " '12499',\n",
       " '13178',\n",
       " '14198',\n",
       " '14450',\n",
       " '14544',\n",
       " '14917',\n",
       " '14918',\n",
       " '14919',\n",
       " '14923',\n",
       " '15452',\n",
       " '16434',\n",
       " '171567',\n",
       " '18102',\n",
       " '18103',\n",
       " '18160',\n",
       " '18573',\n",
       " '18574',\n",
       " '18575',\n",
       " '18576',\n",
       " '18577',\n",
       " '18578',\n",
       " '18582',\n",
       " '18583',\n",
       " '18584',\n",
       " '18585',\n",
       " '18587',\n",
       " '18588',\n",
       " '18605',\n",
       " '18746',\n",
       " '18770',\n",
       " '18950',\n",
       " '18968',\n",
       " '18969',\n",
       " '18971',\n",
       " '18972',\n",
       " '18973',\n",
       " '18974',\n",
       " '19075',\n",
       " '19076',\n",
       " '19139',\n",
       " '20016',\n",
       " '20017',\n",
       " '20018',\n",
       " '20019',\n",
       " '20020',\n",
       " '20021',\n",
       " '20022',\n",
       " '20133',\n",
       " '20135',\n",
       " '207728',\n",
       " '209558',\n",
       " '210044',\n",
       " '215446',\n",
       " '218461',\n",
       " '218832',\n",
       " '22262',\n",
       " '224129',\n",
       " '22436',\n",
       " '224794',\n",
       " '225600',\n",
       " '229363',\n",
       " '229589',\n",
       " '229665',\n",
       " '229949',\n",
       " '230103',\n",
       " '230718',\n",
       " '231327',\n",
       " '231329',\n",
       " '231903',\n",
       " '234889',\n",
       " '237823',\n",
       " '238871',\n",
       " '239134',\n",
       " '23917',\n",
       " '23918',\n",
       " '23959',\n",
       " '23971',\n",
       " '23972',\n",
       " '23984',\n",
       " '241489',\n",
       " '242202',\n",
       " '245650',\n",
       " '245841',\n",
       " '26939',\n",
       " '271639',\n",
       " '27369',\n",
       " '28071',\n",
       " '29863',\n",
       " '328099',\n",
       " '382985',\n",
       " '432530',\n",
       " '50773',\n",
       " '53893',\n",
       " '54195',\n",
       " '54369',\n",
       " '54611',\n",
       " '56248',\n",
       " '56520',\n",
       " '59001',\n",
       " '60596',\n",
       " '633979',\n",
       " '64424',\n",
       " '66136',\n",
       " '66355',\n",
       " '66358',\n",
       " '66401',\n",
       " '66420',\n",
       " '66491',\n",
       " '66566',\n",
       " '66681',\n",
       " '667034',\n",
       " '66763',\n",
       " '66979',\n",
       " '67005',\n",
       " '67054',\n",
       " '67065',\n",
       " '67464',\n",
       " '67486',\n",
       " '67710',\n",
       " '67967',\n",
       " '68106',\n",
       " '68695',\n",
       " '68870',\n",
       " '69241',\n",
       " '69745',\n",
       " '69833',\n",
       " '69870',\n",
       " '69920',\n",
       " '70408',\n",
       " '70428',\n",
       " '70881',\n",
       " '71701',\n",
       " '72090',\n",
       " '72157',\n",
       " '73532',\n",
       " '74167',\n",
       " '74414',\n",
       " '75456',\n",
       " '75686',\n",
       " '76025',\n",
       " '76952',\n",
       " '76974',\n",
       " '78600',\n",
       " '78801',\n",
       " '78929',\n",
       " '79059',\n",
       " '94041',\n",
       " '13077',\n",
       " '17960',\n",
       " '17961',\n",
       " '17962',\n",
       " '22262',\n",
       " '22436',\n",
       " '100862375',\n",
       " '100862456',\n",
       " '102641086',\n",
       " '103149',\n",
       " '103850',\n",
       " '107569',\n",
       " '110074',\n",
       " '12495',\n",
       " '12497',\n",
       " '12499',\n",
       " '13178',\n",
       " '171567',\n",
       " '18102',\n",
       " '18103',\n",
       " '18605',\n",
       " '18950',\n",
       " '18968',\n",
       " '18969',\n",
       " '18971',\n",
       " '18972',\n",
       " '18973',\n",
       " '18974',\n",
       " '19075',\n",
       " '19076',\n",
       " '20016',\n",
       " '20017',\n",
       " '20018',\n",
       " '20019',\n",
       " '20020',\n",
       " '20021',\n",
       " '20022',\n",
       " '20133',\n",
       " '20135',\n",
       " '209558',\n",
       " '215446',\n",
       " '21877',\n",
       " '218832',\n",
       " '21915',\n",
       " '22169',\n",
       " '22171',\n",
       " '22245',\n",
       " '22247',\n",
       " '22271',\n",
       " '230718',\n",
       " '231329',\n",
       " '232223',\n",
       " '23959',\n",
       " '245841',\n",
       " '26462',\n",
       " '26939',\n",
       " '28071',\n",
       " '320685',\n",
       " '331487',\n",
       " '382985',\n",
       " '50493',\n",
       " '50773',\n",
       " '51797',\n",
       " '54369',\n",
       " '55936',\n",
       " '56520',\n",
       " '56749',\n",
       " '57813',\n",
       " '59001',\n",
       " '633979',\n",
       " '64424',\n",
       " '64705',\n",
       " '66136',\n",
       " '66401',\n",
       " '66420',\n",
       " '66422',\n",
       " '66491',\n",
       " '66588',\n",
       " '667034',\n",
       " '66763',\n",
       " '66979',\n",
       " '67005',\n",
       " '67065',\n",
       " '67464',\n",
       " '67486',\n",
       " '67710',\n",
       " '67967',\n",
       " '68106',\n",
       " '68556',\n",
       " '69241',\n",
       " '69719',\n",
       " '69745',\n",
       " '69833',\n",
       " '69870',\n",
       " '69920',\n",
       " '70408',\n",
       " '70428',\n",
       " '70881',\n",
       " '71701',\n",
       " '72090',\n",
       " '72269',\n",
       " '72962',\n",
       " '74414',\n",
       " '76025',\n",
       " '76654',\n",
       " '76952',\n",
       " '78929',\n",
       " '79059',\n",
       " '80914',\n",
       " '99586',\n",
       " '100328588',\n",
       " '100470',\n",
       " '108653',\n",
       " '108682',\n",
       " '109900',\n",
       " '11484',\n",
       " '11564',\n",
       " '11565',\n",
       " '11566',\n",
       " '11611',\n",
       " '11898',\n",
       " '14204',\n",
       " '14415',\n",
       " '14417',\n",
       " '14583',\n",
       " '14584',\n",
       " '14645',\n",
       " '14660',\n",
       " '14661',\n",
       " '14718',\n",
       " '14719',\n",
       " '194237',\n",
       " '212647',\n",
       " '214579',\n",
       " '216456',\n",
       " '227231',\n",
       " '231327',\n",
       " '268782',\n",
       " '268860',\n",
       " '269642',\n",
       " '27053',\n",
       " '52633',\n",
       " '53320',\n",
       " '69719',\n",
       " '70503',\n",
       " '73988',\n",
       " '76282',\n",
       " '100678',\n",
       " '104174',\n",
       " '107272',\n",
       " '107869',\n",
       " '108037',\n",
       " '109731',\n",
       " '110695',\n",
       " '11611',\n",
       " '11655',\n",
       " '11656',\n",
       " '11754',\n",
       " '12116',\n",
       " '12183',\n",
       " '12411',\n",
       " '13142',\n",
       " '13382',\n",
       " '14431',\n",
       " '14711',\n",
       " '17161',\n",
       " '18648',\n",
       " '19193',\n",
       " '192166',\n",
       " '20425',\n",
       " '218865',\n",
       " '231691',\n",
       " '235582',\n",
       " '236539',\n",
       " '237940',\n",
       " '257635',\n",
       " ...]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "data = pd.DataFrame({'Pathway IDs':pathways,\"Genes IDs \":genes})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes IDs</th>\n",
       "      <th>Pathway IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103988</td>\n",
       "      <td>00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>106557</td>\n",
       "      <td>00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110695</td>\n",
       "      <td>00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11522</td>\n",
       "      <td>00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11529</td>\n",
       "      <td>00010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Genes IDs  Pathway IDs\n",
       "0     103988       00010\n",
       "1     106557       00010\n",
       "2     110695       00010\n",
       "3      11522       00010\n",
       "4      11529       00010"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dictionary = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "kegg = {}\n",
    "for i in range(len(data)):\n",
    "    pathway = data.ix[i, 'Pathway IDs']\n",
    "    if pathway not in kegg:\n",
    "        kegg[pathway] = [data.ix[i, 'Genes IDs ']]\n",
    "    else:\n",
    "        kegg[pathway].append(data.ix[i, 'Genes IDs '])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#set(pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['103988',\n",
       " '106557',\n",
       " '110695',\n",
       " '11522',\n",
       " '11529',\n",
       " '11532',\n",
       " '11669',\n",
       " '11670',\n",
       " '11671',\n",
       " '11674',\n",
       " '11676',\n",
       " '12183',\n",
       " '13382',\n",
       " '13806',\n",
       " '13807',\n",
       " '13808',\n",
       " '14120',\n",
       " '14121',\n",
       " '14377',\n",
       " '14378',\n",
       " '14433',\n",
       " '14447',\n",
       " '14751',\n",
       " '15275',\n",
       " '15277',\n",
       " '16828',\n",
       " '16832',\n",
       " '16833',\n",
       " '17330',\n",
       " '18534',\n",
       " '18597',\n",
       " '18598',\n",
       " '18641',\n",
       " '18642',\n",
       " '18648',\n",
       " '18655',\n",
       " '18663',\n",
       " '18746',\n",
       " '18770',\n",
       " '212032',\n",
       " '216019',\n",
       " '21991',\n",
       " '230163',\n",
       " '235339',\n",
       " '26876',\n",
       " '319625',\n",
       " '353204',\n",
       " '433182',\n",
       " '56012',\n",
       " '56421',\n",
       " '56752',\n",
       " '56847',\n",
       " '58810',\n",
       " '60525',\n",
       " '621603',\n",
       " '66681',\n",
       " '67689',\n",
       " '68263',\n",
       " '68401',\n",
       " '68738',\n",
       " '72141',\n",
       " '72157',\n",
       " '72535',\n",
       " '73458',\n",
       " '74551']"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg[\"00010\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pathway_set = set(pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'dict' object has no attribute 'iterkeys'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-39-8db0dbe89456>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mkegg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterkeys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'dict' object has no attribute 'iterkeys'"
     ]
    }
   ],
   "source": [
    "kegg.iterkeys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "Missing parentheses in call to 'print' (<ipython-input-40-799ff59440f1>, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-40-799ff59440f1>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    print pathway\u001b[0m\n\u001b[0m                ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m Missing parentheses in call to 'print'\n"
     ]
    }
   ],
   "source": [
    "for pathway,key in kegg:\n",
    "    print pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'dict' object has no attribute 'iteritems'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-41-370401ed168d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mkegg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miteritems\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'dict' object has no attribute 'iteritems'"
     ]
    }
   ],
   "source": [
    "kegg.iteritems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gmt = ''\n",
    "for key in sorted(kegg.iterkeys()):\n",
    "    for \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Pathway: 00010\\t\\tPathway: 00020\\t\\tPathway: 00030\\t\\tPathway: 00040\\t\\tPathway: 00051\\t\\tPathway: 00052\\t\\tPathway: 00053\\t\\tPathway: 00061\\t\\tPathway: 00062\\t\\tPathway: 00071\\t\\tPathway: 00072\\t\\tPathway: 00100\\t\\tPathway: 00120\\t\\tPathway: 00130\\t\\tPathway: 00140\\t\\tPathway: 00190\\t\\tPathway: 00220\\t\\tPathway: 00230\\t\\tPathway: 00232\\t\\tPathway: 00240\\t\\tPathway: 00250\\t\\tPathway: 00260\\t\\tPathway: 00270\\t\\tPathway: 00280\\t\\tPathway: 00290\\t\\tPathway: 00300\\t\\tPathway: 00310\\t\\tPathway: 00330\\t\\tPathway: 00340\\t\\tPathway: 00350\\t\\tPathway: 00360\\t\\tPathway: 00380\\t\\tPathway: 00400\\t\\tPathway: 00410\\t\\tPathway: 00430\\t\\tPathway: 00450\\t\\tPathway: 00471\\t\\tPathway: 00472\\t\\tPathway: 00480\\t\\tPathway: 00500\\t\\tPathway: 00510\\t\\tPathway: 00511\\t\\tPathway: 00512\\t\\tPathway: 00514\\t\\tPathway: 00520\\t\\tPathway: 00524\\t\\tPathway: 00531\\t\\tPathway: 00532\\t\\tPathway: 00533\\t\\tPathway: 00534\\t\\tPathway: 00561\\t\\tPathway: 00562\\t\\tPathway: 00563\\t\\tPathway: 00564\\t\\tPathway: 00565\\t\\tPathway: 00590\\t\\tPathway: 00591\\t\\tPathway: 00592\\t\\tPathway: 00600\\t\\tPathway: 00601\\t\\tPathway: 00603\\t\\tPathway: 00604\\t\\tPathway: 00620\\t\\tPathway: 00630\\t\\tPathway: 00640\\t\\tPathway: 00650\\t\\tPathway: 00670\\t\\tPathway: 00730\\t\\tPathway: 00740\\t\\tPathway: 00750\\t\\tPathway: 00760\\t\\tPathway: 00770\\t\\tPathway: 00780\\t\\tPathway: 00785\\t\\tPathway: 00790\\t\\tPathway: 00830\\t\\tPathway: 00860\\t\\tPathway: 00900\\t\\tPathway: 00910\\t\\tPathway: 00920\\t\\tPathway: 00970\\t\\tPathway: 00980\\t\\tPathway: 00982\\t\\tPathway: 00983\\t\\tPathway: 01040\\t\\tPathway: 01100\\t\\tPathway: 01200\\t\\tPathway: 01210\\t\\tPathway: 01212\\t\\tPathway: 01230\\t\\tPathway: 01521\\t\\tPathway: 01522\\t\\tPathway: 01524\\t\\tPathway: 02010\\t\\tPathway: 03008\\t\\tPathway: 03010\\t\\tPathway: 03013\\t\\tPathway: 03015\\t\\tPathway: 03018\\t\\tPathway: 03020\\t\\tPathway: 03022\\t\\tPathway: 03030\\t\\tPathway: 03040\\t\\tPathway: 03050\\t\\tPathway: 03060\\t\\tPathway: 03320\\t\\tPathway: 03410\\t\\tPathway: 03420\\t\\tPathway: 03430\\t\\tPathway: 03440\\t\\tPathway: 03450\\t\\tPathway: 03460\\t\\tPathway: 04010\\t\\tPathway: 04012\\t\\tPathway: 04014\\t\\tPathway: 04015\\t\\tPathway: 04020\\t\\tPathway: 04022\\t\\tPathway: 04024\\t\\tPathway: 04060\\t\\tPathway: 04062\\t\\tPathway: 04064\\t\\tPathway: 04066\\t\\tPathway: 04068\\t\\tPathway: 04070\\t\\tPathway: 04071\\t\\tPathway: 04072\\t\\tPathway: 04080\\t\\tPathway: 04110\\t\\tPathway: 04114\\t\\tPathway: 04115\\t\\tPathway: 04120\\t\\tPathway: 04122\\t\\tPathway: 04130\\t\\tPathway: 04140\\t\\tPathway: 04141\\t\\tPathway: 04142\\t\\tPathway: 04144\\t\\tPathway: 04145\\t\\tPathway: 04146\\t\\tPathway: 04150\\t\\tPathway: 04151\\t\\tPathway: 04152\\t\\tPathway: 04210\\t\\tPathway: 04211\\t\\tPathway: 04213\\t\\tPathway: 04215\\t\\tPathway: 04260\\t\\tPathway: 04261\\t\\tPathway: 04270\\t\\tPathway: 04310\\t\\tPathway: 04320\\t\\tPathway: 04330\\t\\tPathway: 04340\\t\\tPathway: 04350\\t\\tPathway: 04360\\t\\tPathway: 04370\\t\\tPathway: 04380\\t\\tPathway: 04390\\t\\tPathway: 04392\\t\\tPathway: 04510\\t\\tPathway: 04512\\t\\tPathway: 04514\\t\\tPathway: 04520\\t\\tPathway: 04530\\t\\tPathway: 04540\\t\\tPathway: 04550\\t\\tPathway: 04610\\t\\tPathway: 04611\\t\\tPathway: 04612\\t\\tPathway: 04614\\t\\tPathway: 04620\\t\\tPathway: 04621\\t\\tPathway: 04622\\t\\tPathway: 04623\\t\\tPathway: 04630\\t\\tPathway: 04640\\t\\tPathway: 04650\\t\\tPathway: 04660\\t\\tPathway: 04662\\t\\tPathway: 04664\\t\\tPathway: 04666\\t\\tPathway: 04668\\t\\tPathway: 04670\\t\\tPathway: 04672\\t\\tPathway: 04710\\t\\tPathway: 04713\\t\\tPathway: 04720\\t\\tPathway: 04721\\t\\tPathway: 04722\\t\\tPathway: 04723\\t\\tPathway: 04724\\t\\tPathway: 04725\\t\\tPathway: 04726\\t\\tPathway: 04727\\t\\tPathway: 04728\\t\\tPathway: 04730\\t\\tPathway: 04740\\t\\tPathway: 04742\\t\\tPathway: 04744\\t\\tPathway: 04750\\t\\tPathway: 04810\\t\\tPathway: 04910\\t\\tPathway: 04911\\t\\tPathway: 04912\\t\\tPathway: 04913\\t\\tPathway: 04914\\t\\tPathway: 04915\\t\\tPathway: 04916\\t\\tPathway: 04917\\t\\tPathway: 04918\\t\\tPathway: 04919\\t\\tPathway: 04920\\t\\tPathway: 04921\\t\\tPathway: 04922\\t\\tPathway: 04923\\t\\tPathway: 04924\\t\\tPathway: 04925\\t\\tPathway: 04930\\t\\tPathway: 04931\\t\\tPathway: 04932\\t\\tPathway: 04933\\t\\tPathway: 04940\\t\\tPathway: 04950\\t\\tPathway: 04960\\t\\tPathway: 04961\\t\\tPathway: 04962\\t\\tPathway: 04964\\t\\tPathway: 04966\\t\\tPathway: 04970\\t\\tPathway: 04971\\t\\tPathway: 04972\\t\\tPathway: 04973\\t\\tPathway: 04974\\t\\tPathway: 04975\\t\\tPathway: 04976\\t\\tPathway: 04977\\t\\tPathway: 04978\\t\\tPathway: 05010\\t\\tPathway: 05012\\t\\tPathway: 05014\\t\\tPathway: 05016\\t\\tPathway: 05020\\t\\tPathway: 05030\\t\\tPathway: 05031\\t\\tPathway: 05032\\t\\tPathway: 05033\\t\\tPathway: 05034\\t\\tPathway: 05100\\t\\tPathway: 05132\\t\\tPathway: 05133\\t\\tPathway: 05134\\t\\tPathway: 05140\\t\\tPathway: 05142\\t\\tPathway: 05143\\t\\tPathway: 05144\\t\\tPathway: 05145\\t\\tPathway: 05146\\t\\tPathway: 05150\\t\\tPathway: 05152\\t\\tPathway: 05160\\t\\tPathway: 05161\\t\\tPathway: 05162\\t\\tPathway: 05164\\t\\tPathway: 05166\\t\\tPathway: 05168\\t\\tPathway: 05169\\t\\tPathway: 05200\\t\\tPathway: 05202\\t\\tPathway: 05203\\t\\tPathway: 05204\\t\\tPathway: 05205\\t\\tPathway: 05206\\t\\tPathway: 05210\\t\\tPathway: 05211\\t\\tPathway: 05212\\t\\tPathway: 05213\\t\\tPathway: 05214\\t\\tPathway: 05215\\t\\tPathway: 05216\\t\\tPathway: 05217\\t\\tPathway: 05218\\t\\tPathway: 05219\\t\\tPathway: 05220\\t\\tPathway: 05221\\t\\tPathway: 05222\\t\\tPathway: 05223\\t\\tPathway: 05224\\t\\tPathway: 05230\\t\\tPathway: 05231\\t\\tPathway: 05310\\t\\tPathway: 05320\\t\\tPathway: 05321\\t\\tPathway: 05322\\t\\tPathway: 05323\\t\\tPathway: 05330\\t\\tPathway: 05332\\t\\tPathway: 05340\\t\\tPathway: 05410\\t\\tPathway: 05412\\t\\tPathway: 05414\\t\\tPathway: 05416\\t\\t'"
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "# *Always* tell NCBI who you are\n",
    "Entrez.email = \"your email here\"\n",
    "\n",
    "def retrieve_annotation(id_list):\n",
    "\n",
    "    \"\"\"Annotates Entrez Gene IDs using Bio.Entrez, in particular epost (to\n",
    "    submit the data to NCBI) and esummary to retrieve the information.\n",
    "    Returns a list of dictionaries with the annotations.\"\"\"\n",
    "\n",
    "    request = Entrez.epost(\"gene\",id=\",\".join(id_list))\n",
    "    try:\n",
    "        result = Entrez.read(request)\n",
    "    except RuntimeError as e:\n",
    "        #FIXME: How generate NAs instead of causing an error with invalid IDs?\n",
    "        print (\"An error occurred while retrieving the annotations.\")\n",
    "        print (\"The error returned was %s\" % e)\n",
    "        sys.exit(-1)\n",
    "\n",
    "    webEnv = result[\"WebEnv\"]\n",
    "    queryKey = result[\"QueryKey\"]\n",
    "    data = Entrez.esummary(db=\"gene\", webenv=webEnv, query_key =\n",
    "            queryKey)\n",
    "    annotations = Entrez.read(data)\n",
    "\n",
    "    print (\"Retrieved %d annotations for %d genes\" % (len(annotations),\n",
    "            len(id_list)))\n",
    "\n",
    "    return annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieved 1 annotations for 6 genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'NomenclatureStatus': 'Official', 'Chromosome': '19', 'GeneticSource': 'genomic', 'Mim': ['138670'], 'ChrSort': '19', 'Summary': 'The protein encoded by this gene is a plasma glycoprotein of unknown function. The protein shows sequence similarity to the variable regions of some immunoglobulin supergene family member proteins. [provided by RefSeq, Jul 2008]', 'NomenclatureName': 'alpha-1-B glycoprotein', 'LocationHist': [{'ChrStart': '58353498', 'ChrStop': '58346805', 'AnnotationRelease': '108', 'AssemblyAccVer': 'GCF_000001405.33', 'ChrAccVer': 'NC_000019.10'}, {'ChrStart': '58858794', 'ChrStop': '58852104', 'AnnotationRelease': '108', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018930.2'}, {'ChrStart': '58353498', 'ChrStop': '58346805', 'AnnotationRelease': '107', 'AssemblyAccVer': 'GCF_000001405.28', 'ChrAccVer': 'NC_000019.10'}, {'ChrStart': '58858794', 'ChrStop': '58852104', 'AnnotationRelease': '107', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018930.2'}, {'ChrStart': '58353498', 'ChrStop': '58346805', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000001405.26', 'ChrAccVer': 'NC_000019.10'}, {'ChrStart': '55174019', 'ChrStop': '55167315', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}, {'ChrStart': '58858794', 'ChrStop': '58852104', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018930.2'}, {'ChrStart': '58864864', 'ChrStop': '58858171', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000001405.25', 'ChrAccVer': 'NC_000019.9'}, {'ChrStart': '55174019', 'ChrStop': '55167315', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}, {'ChrStart': '58858794', 'ChrStop': '58852104', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018930.2'}, {'ChrStart': '58864864', 'ChrStop': '58858171', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000001405.22', 'ChrAccVer': 'NC_000019.9'}, {'ChrStart': '55174019', 'ChrStop': '55167315', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}, {'ChrStart': '59013660', 'ChrStop': '59006970', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000306695.1', 'ChrAccVer': 'NC_018930.1'}, {'ChrStart': '58864864', 'ChrStop': '58858171', 'AnnotationRelease': '103', 'AssemblyAccVer': 'GCF_000001405.21', 'ChrAccVer': 'NC_000019.9'}, {'ChrStart': '55174019', 'ChrStop': '55167315', 'AnnotationRelease': '103', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}, {'ChrStart': '58864864', 'ChrStop': '58858171', 'AnnotationRelease': '37.3', 'AssemblyAccVer': 'GCF_000001405.17', 'ChrAccVer': 'NC_000019.9'}, {'ChrStart': '55174019', 'ChrStop': '55167315', 'AnnotationRelease': '37.3', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}, {'ChrStart': '58864864', 'ChrStop': '58858171', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000001405.14', 'ChrAccVer': 'NC_000019.9'}, {'ChrStart': '55908857', 'ChrStop': '55902159', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000062.1'}, {'ChrStart': '55174019', 'ChrStop': '55167315', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}, {'ChrStart': '58864864', 'ChrStop': '58858171', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000001405.13', 'ChrAccVer': 'NC_000019.9'}, {'ChrStart': '55908857', 'ChrStop': '55902159', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000062.1'}, {'ChrStart': '55174019', 'ChrStop': '55167315', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}, {'ChrStart': '63556668', 'ChrStop': '63548355', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000001405.12', 'ChrAccVer': 'NC_000019.8'}, {'ChrStart': '55908849', 'ChrStop': '55900530', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000062.1'}, {'ChrStart': '55174011', 'ChrStop': '55165686', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000151.1'}], 'Organism': {'CommonName': 'human', 'TaxID': '9606', 'ScientificName': 'Homo sapiens'}, 'MapLocation': '19q13.43', 'OtherDesignations': 'alpha-1B-glycoprotein|HEL-S-163pA|epididymis secretory sperm binding protein Li 163pA', 'Name': 'A1BG', 'OtherAliases': 'A1B, ABG, GAB, HYST2477', 'GeneWeight': '2455', 'GenomicInfo': [{'ChrLoc': '19', 'ChrStop': '58346805', 'ChrStart': '58353498', 'ChrAccVer': 'NC_000019.10', 'ExonCount': '8'}], 'CurrentID': '0', 'NomenclatureSymbol': 'A1BG', 'ChrStart': '58346805', 'Status': '0', 'Description': 'alpha-1-B glycoprotein'}, attributes={'uid': '1'}), DictElement({'NomenclatureStatus': 'Official', 'Chromosome': '12', 'GeneticSource': 'genomic', 'Mim': [], 'ChrSort': '12', 'Summary': '', 'NomenclatureName': 'alpha-2-macroglobulin pseudogene 1', 'LocationHist': [{'ChrStart': '9234206', 'ChrStop': '9228532', 'AnnotationRelease': '108', 'AssemblyAccVer': 'GCF_000001405.33', 'ChrAccVer': 'NC_000012.12'}, {'ChrStart': '9355738', 'ChrStop': '9350067', 'AnnotationRelease': '108', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018923.2'}, {'ChrStart': '9234206', 'ChrStop': '9228532', 'AnnotationRelease': '107', 'AssemblyAccVer': 'GCF_000001405.28', 'ChrAccVer': 'NC_000012.12'}, {'ChrStart': '9355738', 'ChrStop': '9350067', 'AnnotationRelease': '107', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018923.2'}, {'ChrStart': '9234206', 'ChrStop': '9228532', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000001405.26', 'ChrAccVer': 'NC_000012.12'}, {'ChrStart': '9167677', 'ChrStop': '9162006', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}, {'ChrStart': '9355738', 'ChrStop': '9350067', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018923.2'}, {'ChrStart': '9386802', 'ChrStop': '9381128', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000001405.25', 'ChrAccVer': 'NC_000012.11'}, {'ChrStart': '9167677', 'ChrStop': '9162006', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}, {'ChrStart': '9355738', 'ChrStop': '9350067', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018923.2'}, {'ChrStart': '9386802', 'ChrStop': '9381128', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000001405.22', 'ChrAccVer': 'NC_000012.11'}, {'ChrStart': '9167677', 'ChrStop': '9162006', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}, {'ChrStart': '9265422', 'ChrStop': '9259751', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000306695.1', 'ChrAccVer': 'NC_018923.1'}, {'ChrStart': '9386802', 'ChrStop': '9381128', 'AnnotationRelease': '103', 'AssemblyAccVer': 'GCF_000001405.21', 'ChrAccVer': 'NC_000012.11'}, {'ChrStart': '9167677', 'ChrStop': '9162006', 'AnnotationRelease': '103', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}, {'ChrStart': '9386802', 'ChrStop': '9381128', 'AnnotationRelease': '37.3', 'AssemblyAccVer': 'GCF_000001405.17', 'ChrAccVer': 'NC_000012.11'}, {'ChrStart': '9167677', 'ChrStop': '9162006', 'AnnotationRelease': '37.3', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}, {'ChrStart': '9386907', 'ChrStop': '9384201', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000001405.14', 'ChrAccVer': 'NC_000012.11'}, {'ChrStart': '10924047', 'ChrStop': '10921341', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000055.1'}, {'ChrStart': '9167782', 'ChrStop': '9165076', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}, {'ChrStart': '9386907', 'ChrStop': '9384201', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000001405.13', 'ChrAccVer': 'NC_000012.11'}, {'ChrStart': '10924047', 'ChrStop': '10921341', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000055.1'}, {'ChrStart': '9167782', 'ChrStop': '9165076', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}, {'ChrStart': '9278174', 'ChrStop': '9275468', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000001405.12', 'ChrAccVer': 'NC_000012.10'}, {'ChrStart': '10924047', 'ChrStop': '10921341', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000055.1'}, {'ChrStart': '9167782', 'ChrStop': '9165076', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000144.1'}], 'Organism': {'CommonName': 'human', 'TaxID': '9606', 'ScientificName': 'Homo sapiens'}, 'MapLocation': '12p13.31', 'OtherDesignations': 'pregnancy-zone protein pseudogene', 'Name': 'A2MP1', 'OtherAliases': 'A2MP', 'GeneWeight': '689', 'GenomicInfo': [{'ChrLoc': '12', 'ChrStop': '9228532', 'ChrStart': '9234206', 'ChrAccVer': 'NC_000012.12', 'ExonCount': '9'}], 'CurrentID': '0', 'NomenclatureSymbol': 'A2MP1', 'ChrStart': '9228532', 'Status': '0', 'Description': 'alpha-2-macroglobulin pseudogene 1'}, attributes={'uid': '3'}), DictElement({'NomenclatureStatus': 'Official', 'Chromosome': '8', 'GeneticSource': 'genomic', 'Mim': ['108345'], 'ChrSort': '08', 'Summary': 'This gene is one of two arylamine N-acetyltransferase (NAT) genes in the human genome, and is orthologous to the mouse and rat Nat2 genes. The enzyme encoded by this gene catalyzes the transfer of an acetyl group from acetyl-CoA to various arylamine and hydrazine substrates. This enzyme helps metabolize drugs and other xenobiotics, and functions in folate catabolism. Multiple transcript variants encoding different isoforms have been found for this gene. [provided by RefSeq, Aug 2011]', 'NomenclatureName': 'N-acetyltransferase 1', 'LocationHist': [{'ChrStart': '18170418', 'ChrStop': '18223688', 'AnnotationRelease': '108', 'AssemblyAccVer': 'GCF_000001405.33', 'ChrAccVer': 'NC_000008.11'}, {'ChrStart': '18229425', 'ChrStop': '18282650', 'AnnotationRelease': '108', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018919.2'}, {'ChrStart': '18170461', 'ChrStop': '18223688', 'AnnotationRelease': '107', 'AssemblyAccVer': 'GCF_000001405.28', 'ChrAccVer': 'NC_000008.11'}, {'ChrStart': '18229425', 'ChrStop': '18282650', 'AnnotationRelease': '107', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018919.2'}, {'ChrStart': '18170461', 'ChrStop': '18223688', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000001405.26', 'ChrAccVer': 'NC_000008.11'}, {'ChrStart': '16572273', 'ChrStop': '16625519', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}, {'ChrStart': '18229425', 'ChrStop': '18282650', 'AnnotationRelease': '106', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018919.2'}, {'ChrStart': '18027970', 'ChrStop': '18081197', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000001405.25', 'ChrAccVer': 'NC_000008.10'}, {'ChrStart': '16572273', 'ChrStop': '16625519', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}, {'ChrStart': '18229425', 'ChrStop': '18282650', 'AnnotationRelease': '105', 'AssemblyAccVer': 'GCF_000306695.2', 'ChrAccVer': 'NC_018919.2'}, {'ChrStart': '18027970', 'ChrStop': '18081197', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000001405.22', 'ChrAccVer': 'NC_000008.10'}, {'ChrStart': '16572273', 'ChrStop': '16625519', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}, {'ChrStart': '18282846', 'ChrStop': '18336067', 'AnnotationRelease': '104', 'AssemblyAccVer': 'GCF_000306695.1', 'ChrAccVer': 'NC_018919.1'}, {'ChrStart': '18027970', 'ChrStop': '18081197', 'AnnotationRelease': '103', 'AssemblyAccVer': 'GCF_000001405.21', 'ChrAccVer': 'NC_000008.10'}, {'ChrStart': '16572273', 'ChrStop': '16625519', 'AnnotationRelease': '103', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}, {'ChrStart': '18067289', 'ChrStop': '18081197', 'AnnotationRelease': '37.3', 'AssemblyAccVer': 'GCF_000001405.17', 'ChrAccVer': 'NC_000008.10'}, {'ChrStart': '16611618', 'ChrStop': '16625519', 'AnnotationRelease': '37.3', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}, {'ChrStart': '18067289', 'ChrStop': '18081197', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000001405.14', 'ChrAccVer': 'NC_000008.10'}, {'ChrStart': '17033181', 'ChrStop': '17047128', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000051.1'}, {'ChrStart': '16611618', 'ChrStop': '16625519', 'AnnotationRelease': '37.2', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}, {'ChrStart': '18067614', 'ChrStop': '18080819', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000001405.13', 'ChrAccVer': 'NC_000008.10'}, {'ChrStart': '17033506', 'ChrStop': '17046750', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000051.1'}, {'ChrStart': '16611943', 'ChrStop': '16625142', 'AnnotationRelease': '37.1', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}, {'ChrStart': '18111894', 'ChrStop': '18125099', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000001405.12', 'ChrAccVer': 'NC_000008.9'}, {'ChrStart': '17033528', 'ChrStop': '17047128', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000002115.2', 'ChrAccVer': 'AC_000051.1'}, {'ChrStart': '16611943', 'ChrStop': '16625142', 'AnnotationRelease': '36.3', 'AssemblyAccVer': 'GCF_000002125.1', 'ChrAccVer': 'AC_000140.1'}], 'Organism': {'CommonName': 'human', 'TaxID': '9606', 'ScientificName': 'Homo sapiens'}, 'MapLocation': '8p22', 'OtherDesignations': 'arylamine N-acetyltransferase 1|N-acetyltransferase 1 (arylamine N-acetyltransferase)|N-acetyltransferase type 1|arylamide acetylase 1|monomorphic arylamine N-acetyltransferase', 'Name': 'NAT1', 'OtherAliases': 'AAC1, MNAT, NAT-1, NATI', 'GeneWeight': '25480', 'GenomicInfo': [{'ChrLoc': '8', 'ChrStop': '18223688', 'ChrStart': '18170418', 'ChrAccVer': 'NC_000008.11', 'ExonCount': '11'}], 'CurrentID': '0', 'NomenclatureSymbol': 'NAT1', 'ChrStart': '18170418', 'Status': '0', 'Description': 'N-acetyltransferase 1'}, attributes={'uid': '9'}), DictElement({'NomenclatureStatus': 'Official', 'Chromosome': '11', 'GeneticSource': 'genomic', 'Mim': [], 'ChrSort': '~~last', 'Summary': '', 'NomenclatureName': 'atrophia areata, peripapillary chorioretinal degeneration', 'LocationHist': [], 'Organism': {'CommonName': 'human', 'TaxID': '9606', 'ScientificName': 'Homo sapiens'}, 'MapLocation': '11p15', 'OtherDesignations': '', 'Name': 'AA', 'OtherAliases': '', 'GeneWeight': '0', 'GenomicInfo': [], 'CurrentID': '7003', 'NomenclatureSymbol': 'AA', 'ChrStart': '999999999', 'Status': '1', 'Description': 'atrophia areata, peripapillary chorioretinal degeneration'}, attributes={'uid': '8'}), DictElement({'NomenclatureStatus': 'Official', 'Chromosome': '11', 'GeneticSource': 'genomic', 'Mim': [], 'ChrSort': '~~last', 'Summary': '', 'NomenclatureName': 'atrophia areata, peripapillary chorioretinal degeneration', 'LocationHist': [], 'Organism': {'CommonName': 'human', 'TaxID': '9606', 'ScientificName': 'Homo sapiens'}, 'MapLocation': '11p15', 'OtherDesignations': '', 'Name': 'AA', 'OtherAliases': '', 'GeneWeight': '0', 'GenomicInfo': [], 'CurrentID': '7003', 'NomenclatureSymbol': 'AA', 'ChrStart': '999999999', 'Status': '1', 'Description': 'atrophia areata, peripapillary chorioretinal degeneration'}, attributes={'uid': '8'})], 'DbBuild': 'Build161029-0205m.1'}, attributes={'status': 'OK'})}"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "retrieve_annotation(\"103988\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 1.2: Save the KEGG gene sets as a gmt file after you made sure they have the proper gene ids with respect to your DE analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "hints: \n",
    "\n",
    "http://biopython.org/wiki/Annotate_Entrez_Gene_IDs\n",
    "\n",
    "http://www.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2: Gene Set Enrichment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 2.1: Read in the csv file you produced during the Differential Expression module, extract a gene list (as a python list of gene symbols) from your favorite multiple correction column (and store it in a variable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 2.2: Perform gene set enrichment (Fisher's exact test or an hypergeometric test will do for our purposes) with the KEGG gene sets you extracted in Task 1 (you may want to store the results in a pandas dataframe and write them to csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "hint:\n",
    "\n",
    "https://genetrail2.bioinf.uni-sb.de/help?topic=set_level_statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENTRY       103988            CDS       T01002\\nNAME        Gck, GLK, Gk, Gls006, HK4, HKIV, HXKP, Hlb62, MODY2\\nDEFINITION  (RefSeq) glucokinase (EC:2.7.1.2)\\nORTHOLOGY   K12407  glucokinase [EC:2.7.1.2]\\nORGANISM    mmu  Mus musculus (mouse)\\nPATHWAY     mmu00010  Glycolysis / Gluconeogenesis\\n            mmu00052  Galactose metabolism\\n            mmu00500  Starch and sucrose metabolism\\n            mmu00520  Amino sugar and nucleotide sugar metabolism\\n            mmu00524  Neomycin, kanamycin and gentamicin biosynthesis\\n            mmu01100  Metabolic pathways\\n            mmu01200  Carbon metabolism\\n            mmu04910  Insulin signaling pathway\\n            mmu04911  Insulin secretion\\n            mmu04917  Prolactin signaling pathway\\n            mmu04922  Glucagon signaling pathway\\n            mmu04930  Type II diabetes mellitus\\n            mmu04950  Maturity onset diabetes of the young\\n            mmu05230  Central carbon metabolism in cancer\\nMODULE      mmu_M00001  Glycolysis (Embden-Meyerhof pathway), glucose => pyruvate\\n            mmu_M00549  Nucleotide sugar biosynthesis, glucose => UDP-glucose\\nBRITE       KEGG Orthology (KO) [BR:mmu00001]\\n             Metabolism\\n              Overview\\n               01200 Carbon metabolism\\n                103988 (Gck)\\n              Carbohydrate metabolism\\n               00010 Glycolysis / Gluconeogenesis\\n                103988 (Gck)\\n               00052 Galactose metabolism\\n                103988 (Gck)\\n               00500 Starch and sucrose metabolism\\n                103988 (Gck)\\n               00520 Amino sugar and nucleotide sugar metabolism\\n                103988 (Gck)\\n              Biosynthesis of other secondary metabolites\\n               00524 Neomycin, kanamycin and gentamicin biosynthesis\\n                103988 (Gck)\\n             Organismal Systems\\n              Endocrine system\\n               04911 Insulin secretion\\n                103988 (Gck)\\n               04910 Insulin signaling pathway\\n                103988 (Gck)\\n               04922 Glucagon signaling pathway\\n                103988 (Gck)\\n               04917 Prolactin signaling pathway\\n                103988 (Gck)\\n             Human Diseases\\n              Cancers\\n               05230 Central carbon metabolism in cancer\\n                103988 (Gck)\\n              Endocrine and metabolic diseases\\n               04930 Type II diabetes mellitus\\n                103988 (Gck)\\n               04950 Maturity onset diabetes of the young\\n                103988 (Gck)\\n            Enzymes [BR:mmu01000]\\n             2. Transferases\\n              2.7  Transferring phosphorus-containing groups\\n               2.7.1  Phosphotransferases with an alcohol group as acceptor\\n                2.7.1.2  glucokinase\\n                 103988 (Gck)\\nPOSITION    11 A1; 11 3.88 cM\\nMOTIF       Pfam: Hexokinase_2 Hexokinase_1\\nDBLINKS     NCBI-ProteinID: NP_001274315\\n            NCBI-GeneID: 103988\\n            MGI: 1270854\\n            Ensembl: ENSMUSG00000041798\\n            Vega: OTTMUSG00000006327\\n            UniProt: P52792 Q5SVI6\\nAASEQ       465\\n            MAVDTTRRGAQSLTLVEQILAEFQLQEEDLKKVMSRMQKEMDRGLKLETHQEASVKMLPT\\n            YVRSTPEGSEVGDFLSLDLGGTNFRVMLVKVGEGEAGQWSVKTKHQMYSIPEDAMTGTAE\\n            MLFDYISECISDFLDKHQMKHKKLPLGFTFSFPVRHEDIDKGILLNWTKGFKASGAEGNN\\n            IVGLLRDAIKRRGDFEMDVVAMVNDTVATMISCYYEDRQCEVGMIVGTGCNACYMEEMQN\\n            VELVEGDEGRMCVNTEWGAFGNSGELDEFLLEYDRMVDESSVNPGQQLYEKIIGGKYMGE\\n            LVRLVLLKLVEENLLFHGEASEQLRTRGAFETRFVSQVESDSGDRRQILNILSTLGLRPS\\n            VADCDIVRRACESVSTRAAHMCSAGLAGVINRMRESRSEDVMRITVGVDGSVYKLHPSFK\\n            ERFHASVRRLTPNCEITFIESEEGSGRGAALVSAVACKKACMLGQ\\nNTSEQ       1398\\n            atggctgtggatactacaaggaggggagcccagtcgttgactctggtagagcagatcctg\\n            gcagagttccagctgcaggaggaagacctgaagaaggtgatgagccggatgcagaaggag\\n            atggaccgtggcctgaagctggagacccatcaggaggccagtgtaaagatgttgcccacc\\n            tacgtgcgttccaccccagaaggctcagaagttggagactttctctccttagacctggga\\n            ggaaccaacttcagggtgatgctggtgaaagtgggggagggggaggcaggacagtggagc\\n            gtgaagacgaaacaccagatgtattccatccccgaggacgccatgacgggcactgcggag\\n            atgctctttgactacatctctgagtgcatctctgacttcctggacaagcatcagatgaaa\\n            cacaagaaactacccctgggcttcaccttctccttccctgtaaggcacgaagacatagac\\n            aagggcatcctgctcaactggaccaagggcttcaaggcctccggagcagaagggaacaac\\n            atcgtgggacttctccgagatgctatcaagaggagaggggactttgagatggatgtggtg\\n            gcaatggtgaatgacacggtggccacaatgatctcctgctactatgaagaccgccaatgt\\n            gaggtcggcatgattgtgggcaccggctgcaacgcctgctacatggaggagatgcagaat\\n            gtggagctggtggaaggcgatgaggggcgcatgtgtgtcaacacagagtggggcgccttc\\n            gggaactccggtgagctggacgagttcctcctggagtacgaccggatggtggatgagagc\\n            tcagtgaaccccggtcagcagctgtacgaaaagatcattggcggaaagtacatgggcgag\\n            ctggtacgacttgtgctgctcaagctggtagaggagaatcttctgttccacggagaggcc\\n            tcagagcagctgcgcacacgtggtgcttttgagacccgttttgtgtcgcaggtggagagc\\n            gactctggggaccgaaggcagatccttaacatcctgagcactctgggccttcgaccctct\\n            gtcgccgactgcgacattgtgcgccgtgcctgtgaaagcgtgtccactcgcgccgcccac\\n            atgtgctcagcaggactagcgggggtcataaatcgcatgcgcgaaagccgcagtgaggac\\n            gtgatgcgcatcacggtgggcgtggatggctccgtgtacaagctgcacccgagcttcaag\\n            gagcggtttcacgccagtgtgcgcaggctgacacccaactgcgaaatcaccttcattgaa\\n            tcagaggagggcagcggcaggggagccgcactggtctctgcggtggcctgcaagaaggct\\n            tgcatgctgggccagtga\\n///\\n'"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_get(\"mmu:103988\").read()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 2.3: Extract a list of significantly (at 0.05 significance) enriched KEGG pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3: KEGG map visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### hint:\n",
    "\n",
    "http://nbviewer.jupyter.org/github/widdowquinn/notebooks/blob/master/Biopython_KGML_intro.ipynb\n",
    "\n",
    "#### remark:\n",
    "\n",
    "In real life you may want to use the R-based tool pathview: https://bioconductor.org/packages/release/bioc/html/pathview.html (if you insist you can also try to use r2py for using pathview from Python during the practical)\n",
    "\n",
    "For Python (in addition to the Biopyhton module) https://github.com/idekerlab/py2cytoscape in combination with https://github.com/idekerlab/KEGGscape may be another alternative (in the future)\n",
    "\n",
    "Generally speaking, it is always a good idea to pay attention also to other pathway databases like Reactome or WikiPathways ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 3.1: Pick some significantly enriched KEGG pathways of your choice from 2.3 and visualize them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 3.2: Define a a suitable binary color scheme respresenting the fact whether a gene is significantly expressed or not"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "hint: \n",
    "\n",
    "http://www.rapidtables.com/web/color/RGB_Color.htm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 3.3: Visualize the pathway(s) from 3.1 in such a way that the included genes have the corresponding color from 3.2 ( you may need to define a suitable mapping from single genes to what is actually shown in the pathway map...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 3.4: Define a suitable continuous color range representing the log2 fold changes of the all the genes in your data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "hint:\n",
    "\n",
    "http://bsou.io/posts/color-gradients-with-python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subtask 3.5: Visualize the pathway(s) from 3.1 in such a way that the included genes have the corresponding color from 3.4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
